File: cpp/source/Detector.cpp
#include "Detector.h"

Detector::Detector(const char* labelsFile, const char* modelPath, bool cudaEnabled, int imgW, int imgH,
	const float scoreThresh, const float nmsThresh, const int maxDet)
	: m_labelsFile{ labelsFile }, m_modelPath{ modelPath }, m_cudaEnabled{ cudaEnabled }, m_imgW{ imgW }, m_imgH{ imgH },
	m_scoreThresh{ scoreThresh }, m_nmsThresh{ nmsThresh }, m_maxDet{ maxDet },
	onnxInferencer{ OnnxInference(m_modelPath, m_labels, m_imgW, m_imgH, m_cudaEnabled, m_scoreThresh, m_nmsThresh, m_maxDet) },
	torchInferencer{ TorchInference(m_modelPath, m_labels, m_imgW, m_imgH, m_cudaEnabled, m_scoreThresh, m_nmsThresh, m_maxDet) } {
	std::ifstream file(Detector::m_labelsFile);

	if (file.is_open()) {
		std::string line;

		while (getline(file, line)) {
			std::stringstream ss(line);
			int i = 0;
			int label;
			std::string labelName;
			while (ss.good()) {
				std::string substr;
				getline(ss, substr, ',');
				switch (i) {
				case 0:
					label = stoi(substr);
					break;
				case 1:
					labelName = substr;
					break;
				default:
					break;
				}

				++i;
			}

			m_labels[label] = labelName;
		}
	}

	file.close();

	std::string mPath = m_modelPath;
	if (mPath.substr(mPath.find_last_of(".") + 1) == "onnx") {
		m_isOnnxModel = true;
		onnxInferencer.loadOnnxNetwork();
		inferencer = &onnxInferencer;
	}
	else {
		m_isOnnxModel = false;
		torchInferencer.loadTorchNetwork();
		inferencer = &torchInferencer;
	}
}


cv::Mat Detector::readImage(const char* inputFile) {
	cv::Mat image = cv::imread(inputFile, cv::IMREAD_COLOR);
	if (image.empty()) {
		std::cout << "Image File: " << inputFile << " Is Not Found" << std::endl;
	}

	return image;
}

std::vector<Detection> Detector::detect(const cv::Mat& image) {
	return inferencer->runInference(image);
}

File: python/source/Open_VO/detector.py
import cv2
import numpy as np
from typing import List, Dict, Optional
from .data_types import Detection
from .inferencers import BaseInferencer
from .inferencers import OnnxInferencerPy
from .inferencers import TorchInferencerPy


class DetectorPy:
    def __init__(self, labels_file: str, model_path: str,
                 cuda_enabled: bool = True, img_w: int = 640, img_h: int = 640,
                 score_thresh: float = 0.45, nms_thresh: float = 0.50, max_det: int = 100):
        self.labels_file = labels_file
        self.model_path = model_path
        self.labels: Dict[int, str] = self._load_labels()

        self.is_onnx_model = model_path.lower().endswith(".onnx")
        self.inferencer: Optional[BaseInferencer] = None

        if self.is_onnx_model:
            self.inferencer = OnnxInferencerPy(
                onnx_model_path=model_path, model_classes=self.labels,
                img_w=img_w, img_h=img_h, run_with_cuda=cuda_enabled,
                score_thresh=score_thresh, nms_thresh=nms_thresh, max_det=max_det
            )
        else:
            self.inferencer = TorchInferencerPy(
                torch_model_path=model_path, model_classes=self.labels,
                img_w=img_w, img_h=img_h, run_with_cuda=cuda_enabled,
                score_thresh=score_thresh, nms_thresh=nms_thresh, max_det=max_det
            )
            # TorchInferencerPy calls load_torch_network in its __init__

    def _load_labels(self) -> Dict[int, str]:
        labels_map: Dict[int, str] = {}
        try:
            with open(self.labels_file, 'r') as file:
                for line in file:
                    line = line.strip()
                    if not line: continue
                    parts = line.split(',', 1)  # Split only on the first comma
                    if len(parts) == 2:
                        try:
                            label_id = int(parts[0])
                            label_name = parts[1].strip()
                            labels_map[label_id] = label_name
                        except ValueError:
                            print(f"Warning: Could not parse label line: {line}")
                    else:
                        print(f"Warning: Malformed label line (expected id,name): {line}")
        except IOError as e:
            print(f"Error reading labels file {self.labels_file}: {e}")
        return labels_map

    def read_image(self, image_path: str) -> Optional[np.ndarray]:
        image = cv2.imread(image_path, cv2.IMREAD_COLOR)
        if image is None:
            print(f"Image File: {image_path} Is Not Found or empty.")
            return None
        return image

    def detect(self, image: np.ndarray) -> List[Detection]:
        if self.inferencer is None:
            print("Inferencer not initialized.")
            return []
        if image is None:
            print("Input image to detect is None.")
            return []
        return self.inferencer.run_inference(image)
